custom_components/envisalink_new/alarm_control_panel.py

```diff
@@ -153,7 +149,7 @@ def __init__(
     @property
     def code_format(self) -> CodeFormat | None:
         """Regex for code format or None if no code is required."""
-        if self._code:
+        if self.state == STATE_ALARM_DISARMED and not self._attr_code_arm_required:
```

Comment:
Ugh.  I think this is complicated not by the new info (see comment above) that the DSC night mode requires a code be entered.  So some modes of arming on a DSC require a code and others don't.  It's probably still safe to set `code_arm_required=false` for DSC systems but maybe the keypad should always be displayed (unless a `code` is configured) since you can't know which mode might be armed.

Reply:
You are right tho `*9` does a "no-entry arm" that I never used personally. And it does require a code.

What I don't get tho, is from the ha integration, I don't have the possibility to arm this way (and would disable this either way), so I don't need the keyboard for any reason as I can't do it right now...

Do you mean you'd modify the implementation and give DSC users the ability to "night arm" from the integration by sending `*9 + code`? If this is what you mean, I'd definitely do this under a configuration option personally... And then we can check this option is set and always display the kbd.

Reply:
The integration doesn't explicit use the `*9` keystrokes but uses the `032` TPI command which is the equivalent.

Which card are you using for your alarm?  The stock alarm control card allows you to select which modes you want to display.  I personally only have the away and stay boxes checked but it is possible to enable the night mode as well.

Honeywell requires a code to arm all modes and DSC requires a code to arm Night mode.  There is already provision to provide the code in config which can be used to drive the `code_arm_required` attribute.  So I wonder if instead there should be a config option to control whether a code is required to DISARM?  That would allow one to specify a code but not have it used automatically for disarming the system.

On DSC systems, if one wanted to arm all modes (including night mode) without a code, they could specify the code in config and set it such that a code is required to disarm (`code_disarm_required-true`).  That would allow a no-code arm but still require a code to disarm.

On Honeywell systems, if one didn't want a code to arm, they set the code in config and set `code_disarm_required=true`.  The current behaviour could be preserved with a smart default if the config doesn't exist - which I think would be `code_disarm_required=not code`.

I think I need to walk through all the scenarios a little more carefully though to convince myself this isn't yet another possibility with an edge case that doesn't work.  😬 

I potential side benefit of this could also be that it might be able to enable the zone bypass switches for DSC systems that require a code a zone since it would be possible to specify a code in the config but not automatically use it for disarming.  But I'm probably getting ahead of myself...

Reply:
I've almost convinced myself that a `code_disarm_required` setting could be the solution here.  The current behaviour is this:

```
Panel Type | Is code set? | Keypad on Arm | Keypad on Disarm | Night mode works?
--------------------------------------------------------------------------------
DSC        | yes          | no            | no               | yes
DSC        | no           | yes           | yes              | yes
HONEYWELL  | yes          | no            | no               | yes
HONEYWELL  | no           | yes           | yes              | yes
```

Where "Keypad on Arm" / "Keypad on Disarm" indicate whether the keypad is shown when arming/disarming the system.

And if a `code_disarm_required` config setting is added and the `code_format` function updated appropriately, the behaviour could be this:

```
Panel Type | Is code set? | code_disarm_required | Keypad on Arm | Keypad on Disarm | Night mode works?
------------------------------------------------------------------------------------------------ ------
DSC        | yes          | yes                  | no            | yes              | yes
DSC        | yes          | no                   | no            | no               | yes
DSC        | no           | yes                  | no            | yes              | no
DSC        | no           | no                   | no            | yes              | no
HONEYWELL  | yes          | yes                  | no            | yes              | yes
HONEYWELL  | yes          | no                   | no            | no               | yes
HONEYWELL  | no           | yes                  | yes           | yes              | yes
HONEYWELL  | no           | no                   | yes           | yes              | yes
```

With the new system, the only variant that would no longer be available would be on a DSC system where the keypad is shown for arming the system.  In a practical sense, I think this is only relevant to the Night mode because the Stay and Away modes never require a code so it doesn't matter what someone enters on the keypad in that case.

What do you think?

Reply:
Actually, there's one other thing I need to confirm.  It looks like the Stay and Away quick arm keys can be enabled or disabled on the panel itself.  I'll try and disable them on my system to see what happens.  It's possible that's just not a supported setup and arming will start failing via this integration or it's possible that the EVL/panel will response to the arm command with a request for a code.  If that latter is the case, then the table above may need to be amended because there would be situations where a DSC setup would require a code for arming.

Reply:
> Actually, there's one other thing I need to confirm. It looks like the Stay and Away quick arm keys can be enabled or disabled on the panel itself.

I tested it on my DSC PC1832NK, and your guess was right: if you disable section 015 option 3 (aka. the Quick Arming feature will be {en,dis}abled), when HA tries to arm without a code, the System will reject it. It puts mine in a weird state too for many seconds, where my keyboard becomes unavailable with a "System Unavailable" error on screen.

So yeah :/ yet another thing to consider.

> What do you think?

THB, I would keep it simple. 
 - I think the fact the keyboard disappear is a nice to have. We can drop it, or add an advanced feature option. If a user checks it but the alarm does not support it, drop it.
 -  I think the user expects to see a panel in HA that does the basics. By default DSC systems can quick arm stay, and quick arm away.
 - Everything else is IMO advanced features, and the integration should make them available thru config options. It's the responsibility of the user to match their setup.

If it's possible to query how the alarm is configured thru the API, well then it opens a lot of possibilities, but it's also a lot more work too. And there might be some cases where it is more misleading than anything. There are cases where a clear error message to the user is better than trying too hard. We can even put in the documentation typical errors and how to configure the panel or the integration to enable it.

Reply:
Hey.. just FYI I haven't forgotten about this.  I started having another look at it today and stumbled across [this](https://github.com/home-assistant/core/pull/112540) which introduced a default code to ALL `alarm_control_panel` entities which makes the `code` configuration item redundant and somewhat incompatible.  There was also some interesting commentary related to the stock `envisalink` integration [here](https://github.com/home-assistant/core/issues/118668). 

Hopefully you're okay using your version for the time being while I figure out what to do here.  I appreciate all your input on this.

Reply:
Yeah I'm using a local patch that's fine. And I'm also waiting for a final consensus before trying another version of this PR. Because right now I'm tempted to add the options I proposed earlier, with backward configuration compatibility in mind, but I'll do it only if we agree it's the way to go.

Reply:
About your link, this is actually why I am using your HACS, when I saw an issue like this closed with someone clearly stating the regression I had: https://github.com/home-assistant/core/issues/118668#issuecomment-2231784660, I knew I should not hold my breath until they fix it :D

Reply:
I suspect it won't get fixed there unless someone from the community submits a fix as there isn't anyone, to my knowledge, maintaining the core `envisalink` integration anymore.

Regarding the changes that they made, it seems the direction they've decided is to lift management of a default code up into the alarm control panel entity itself and exposed through the stock UI.  So having the code in this integration (or any alarm integration) seems redundant now.  I started experimenting with what to do about this.  At the moment I have it such that it will use the default code in the alarm control panel entity if it is set and use the code in `envisalink_new`'s config if it isn't set.  And some warnings in the logs for people to set the code in the alarm control panel entity and remove it from this integration.  And then maybe over time eventually remove the ability to set the code in this integration.  

I don't know whether it's strictly necessary to resolve that before addressing your changes here but it feels like they  may intersect.